File: CSharp_projects/AtmSystemModel/src/project5/Infrastructure/project5.Infrastructure.DataAccess/DataBaseRepositories/AccountRepository.cs
using Lab5.Application.Abstractions.Repositories;
using Lab5.Application.Models.Accounts;
using Npgsql;

namespace Lab5.Infrastructure.DataAccess.DataBaseRepositories;

public class AccountRepository : IAccountRepository
{
    private readonly string _connectionString;

    public AccountRepository(string connectionString)
    {
        _connectionString = connectionString;
    }

    public void CreateAccount(int number, string password)
    {
        using var connection = new NpgsqlConnection(_connectionString);
        connection.Open();
        var command = new NpgsqlCommand("INSERT INTO accounts (account_id, account_number, account_pin, account_balance) VALUES (@id, @number, @pin, @balance)", connection);
        command.Parameters.AddWithValue("@number", number);
        command.Parameters.AddWithValue("@pin", password);
        command.Parameters.AddWithValue("@balance", 0);
        command.ExecuteNonQuery();
    }

    public void UpdateAccount(Account account)
    {
        using var connection = new NpgsqlConnection(_connectionString);
        connection.Open();
        var command = new NpgsqlCommand("UPDATE accounts SET account_balance = @balance WHERE account_number = @number", connection);
        command.Parameters.AddWithValue("@balance", account.Balance);
        command.Parameters.AddWithValue("@number", account.Number);
        command.ExecuteNonQuery();
    }

    public Account? FindAccountByNumber(long number)
    {
        var connection = new NpgsqlConnection(_connectionString);
        connection.Open();

        var command = new NpgsqlCommand("select * from accounts where number = @number", connection);

        command.Parameters.AddWithValue("@number", number);

        using NpgsqlDataReader reader = command.ExecuteReader();

        if (reader.Read() is false)
        {
            return null;
        }

        return new Account(reader.GetInt32(1), reader.GetString(2));
    }
}

File: Java_projects/project1/src/main/java/com/project1/repositories/AccountRepository.java
package com.project1.repositories;

import com.project1.models.Account;

import java.util.ArrayList;
import java.util.List;

/**
 * Реализация репозитория аккаунтов с использованием локального хранилища.
 * Этот класс предоставляет методы для поиска, создания и обновления аккаунтов.
 */
public class AccountRepository implements IAccountRepository {
    private List<Account> accounts;

    /**
     * Конструктор, инициализирующий пустой список аккаунтов.
     */
    public AccountRepository() {
        accounts = new ArrayList<>();
    }

    /**
     * Находит аккаунт по номеру.
     *
     * @param number номер аккаунта
     * @return найденный аккаунт или null, если аккаунт с таким номером не найден
     */
    @Override
    public Account findAccountByNumber(long number) {
        for (Account account : accounts) {
            if (account.getNumber() == number) {
                return account;
            }
        }
        return null;
    }

    /**
     * Создает новый аккаунт, если аккаунт с указанным номером еще не существует.
     *
     * @param number  номер нового аккаунта
     * @param password пароль для нового аккаунта
     */
    @Override
    public void createAccount(int number, String password) {
        if (findAccountByNumber(number) == null) {
            accounts.add(new Account(number, password));
        }
    }

    /**
     * Обновляет информацию о существующем аккаунте.
     * Если аккаунт с таким номером найден, обновляется его баланс.
     *
     * @param account объект аккаунта с обновленной информацией
     */
    @Override
    public void updateAccount(Account account) {
        Account accountByNumber = findAccountByNumber(account.getNumber());
        if (accountByNumber != null) {
            accountByNumber.setBalance(account.getBalance());
        }
    }
}
